spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,310 @@
+    this.sparkContext = sparkContext;
+    this.table = table;
+    this.caseSensitive = caseSensitive;
+    this.expectedSchema = expectedSchema;
+    this.localityPreferred = Spark3Util.isLocalityEnabled(table.io(), table.location(), options);
+
+    long tableSplitSize = PropertyUtil.propertyAsLong(table.properties(), SPLIT_SIZE, SPLIT_SIZE_DEFAULT);
+    this.splitSize = Spark3Util.propertyAsLong(options, SparkReadOptions.SPLIT_SIZE, tableSplitSize);
+
+    int tableSplitLookback = PropertyUtil.propertyAsInt(table.properties(), SPLIT_LOOKBACK, SPLIT_LOOKBACK_DEFAULT);
+    this.splitLookback = Spark3Util.propertyAsInt(options, SparkReadOptions.LOOKBACK, tableSplitLookback);
+
+    long tableSplitOpenFileCost = PropertyUtil.propertyAsLong(
+        table.properties(), SPLIT_OPEN_FILE_COST, SPLIT_OPEN_FILE_COST_DEFAULT);
+    this.splitOpenFileCost = Spark3Util.propertyAsLong(options, SPLIT_OPEN_FILE_COST, tableSplitOpenFileCost);
+
+    InitialOffsetStore initialOffsetStore = InitialOffsetStore.getInstance(table, checkpointLocation);
+    this.initialOffset = getOrWriteInitialOffset(initialOffsetStore);
+  }
+
+  @Override
+  public Offset latestOffset() {
+    table.refresh();
+    Snapshot latestSnapshot = table.currentSnapshot();
+    if (latestSnapshot == null) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    // a readStream on an Iceberg table can be started from 2 types of snapshots
+    // 1. a valid starting Snapshot:
+    //      when this valid starting Snapshot is the initialOffset - then, scanAllFiles must be set to true;
+    //      for all StreamingOffsets following this - scanAllFiles must be set to false
+    // 2. START_OFFSET:
+    //      if the stream started on the table from START_OFFSET - it implies - that all the subsequent Snapshots added
+    //      will have all files as net New manifests & hence scanAllFiles can be false.
+    boolean scanAllFiles = !StreamingOffset.START_OFFSET.equals(initialOffset) &&
+        latestSnapshot.snapshotId() == initialOffset.snapshotId();
+
+    long filesNewlyAddedInLatestSnapshot = Iterables.size(latestSnapshot.addedFiles());
+    long existingFilesInheritedByLatestSnapshot = SnapshotUtil.existingDataFiles(latestSnapshot);
+    long positionValue = scanAllFiles ? existingFilesInheritedByLatestSnapshot + filesNewlyAddedInLatestSnapshot :
+        filesNewlyAddedInLatestSnapshot;
+
+    return new StreamingOffset(latestSnapshot.snapshotId(), positionValue, scanAllFiles);
+  }
+
+  @Override
+  public InputPartition[] planInputPartitions(Offset start, Offset end) {
+    if (end.equals(StreamingOffset.START_OFFSET)) {
+      return new InputPartition[0];
+    }
+
+    // broadcast the table metadata as input partitions will be sent to executors
+    Broadcast<Table> tableBroadcast = sparkContext.broadcast(SerializableTable.copyOf(table));
+    String expectedSchemaString = SchemaParser.toJson(expectedSchema);
+
+    Preconditions.checkState(
```

Comment:
I don't think that `IllegalStateException` is appropriate here because the problem isn't with the state of the stream source, it is with the offset passed in. I would probably use `IllegalArgumentException` instead.